### PR TITLE
Fix YPR rotation issues in GSD calculation

### PR DESCRIPTION
**Pull request by daniel-riehm**
- Fixed GSD calculation - now passes reference test case exactly
- In process of doing above, modified conversions between `kwiver::vital::rotation_< T >` and the yaw-pitch-roll system. Yaw, pitch, and roll are conventionally understood to take place in a NED coordinate sytem, where the x axis is pointing north, y is pointing east, and z is pointing down. However, general computer vision practice when processing aerial imagery is to use the ENU system, where x is pointing east. y is north, and z is up. Previously, the `rotation_< T >` class had treated these systems inconsistently. Combining two YPR-derived `rotation_< T >`s would not produce the correct YPR output, and this was one of the things breaking our computation of GSD. It seems previous developers realized the problem and chose to create a workaround utility function `compose_rotations()`, manually writing the YPR rotation calculations independently of `rotation_< T >` or Eigen. This branch unifies YPR  with `rotation_< T >`, removing `compose_rotations()` and adding utility functions to convert `rotation_< T >`s between NED and ENU.
- **Note**: fixing `rotation_< T >`'s treatment of YPR does change the behavior of the YPR conversion functions - the constructor used to convert them to ENU, but now stores them in their native NED system, with the NED-to-ENU conversion moved to an optional external utility function. I think I found and fixed all references inside KWIVER, but this could potentially break external code which uses KWIVER's `rotation_< T >` for YPR, if any such code exists.

### CODE REVIEW

**Comment by daniel-riehm** [comment]
jenkins test this please

**Comment by daniel-riehm** [comment]
jenkins test this please

**Review by russelldj — APPROVED** [review]
Assuming the test succeeds, this seems reasonable. I didn't give the math a huge amount of attention but your description and the fact that it matches the reference approach gives me a lot of confidence.

File: arrows/core/derive_metadata.cxx
Comment (russelldj) [inline arrows/core/derive_metadata.cxx:237]: Ack

File: arrows/core/tests/test_derive_metadata.cxx
Comment (russelldj) [inline arrows/core/tests/test_derive_metadata.cxx:63]: Also probably my fault.

File: vital/types/rotation.cxx
Comment (russelldj) [inline vital/types/rotation.cxx:53]:
```suggestion
  T const half_x = T{ 0.5 } * roll;
```
If you want to prohibit narrowing casts

File: arrows/core/tests/test_derive_metadata.cxx
Comment (mwoehlke-kitware) [reply arrows/core/tests/test_derive_metadata.cxx:63]: Heh, I was wondering about that; 3:2 isn't a bonkers aspect ratio though...

File: vital/types/sfm_constraints.cxx
Comment (mwoehlke-kitware) [inline vital/types/sfm_constraints.cxx:234]:
```suggestion
    R_loc = rotation_d{ platform_heading, platform_pitch, platform_roll } *
            rotation_d{ sensor_rel_az, sensor_rel_el, sensor_rel_roll };
```

(Bunch of others like this...)

**Review by mwoehlke-kitware — COMMENTED** [review]
I've only glanced at this, but this looks like it really needs a careful review of the apparent functional changes particularly in `rotation` (if David hasn't done that already). I may or may not get to it today.

File: vital/types/rotation.cxx
Comment (daniel-riehm) [reply vital/types/rotation.cxx:53]: In this case I don't think we do? Consider when `T` is `float` - we would want the `0.5` to be narrowed to `float`, right?

Comment (mwoehlke-kitware) [reply vital/types/rotation.cxx:53]: Actually, this is probably better written:
```suggestion
  T const half_x = static_cast< T >( 0.5 ) * roll;
```
(Likewise below...)

File: vital/types/rotation.cxx
Comment (mwoehlke-kitware) [inline vital/types/rotation.cxx:156]: Pretty sure there are conversion warnings hiding in here?

```suggestion
  constexpr auto t1 = static_cast< T >( 1.0 );
  constexpr auto t2 = static_cast< T >( 2.0 );
  roll = std::atan2( t2 * ( q_.w() * q_.x() + q_.y() * q_.z() ),
                     t1 - t2 * ( q_.x() * q_.x() + q_.y() * q_.y() ) );
  pitch = std::asin( t2 * ( q_.w() * q_.y() - q_.x() * q_.z() ) );
  yaw = std::atan2( t2 * ( q_.w() * q_.z() + q_.x() * q_.y() ),
                    t1 - t2 * ( q_.y() * q_.y() + q_.z() * q_.z() ) );
```
(Other approaches are possible, this is just the first thing that came to mind. Alternatively, `constexpr T t1 = 1;` is another option along the lines of the above.)

**Review by mwoehlke-kitware — COMMENTED** [review]
(Same disclaimer as previous.)

File: vital/types/rotation.cxx
Comment (daniel-riehm) [reply vital/types/rotation.cxx:156]: Can you explain why `static_cast` is preferred in this case?

File: vital/types/rotation.cxx
Comment (russelldj) [reply vital/types/rotation.cxx:53]: I think my concern was if `T` was not a `double` or `float` and instead was an integer type. But that's probably pedantic and wouldn't happen in practice.

File: vital/types/sfm_constraints.cxx
Comment (russelldj) [inline vital/types/sfm_constraints.cxx:233]: I'm curious if leaving this compose rotations function would be useful, even if it gets renamed to something more precise. Writing this boilerplate seems like it could be prone to errors.

File: vital/types/rotation.cxx
Comment (russelldj) [inline vital/types/rotation.cxx:321]: As mentioned elsewhere, It seems like it could be nice to keep this utility function to avoid writing boilerplate elsewhere.

Comment (russelldj) [reply vital/types/rotation.cxx:321]: The one which takes rotations as input seems like it could be easily removed though.

File: vital/types/rotation.cxx
Comment (russelldj) [inline vital/types/rotation.cxx:244]: This seems reasonable but I haven't gotten a chance to dig into the constructor for this.

File: vital/types/rotation.cxx
Comment (russelldj) [inline vital/types/rotation.cxx:152]: Is this supposed to produce equivalent results or has there been a change of convention for the input data?

Comment (daniel-riehm) [reply vital/types/rotation.cxx:152]: The object's internal representation has changed convention, so `this` is different. I also switch from the `quaternion -> matrix -> ypr` method shown previously to a more direct / efficient `quaternion -> ypr` method.

Comment (russelldj) [reply vital/types/rotation.cxx:152]: I see. So in both cases we're just converting to/from NED YPR to quaternions?

Comment (daniel-riehm) [reply vital/types/rotation.cxx]: Yes.

File: vital/types/rotation.cxx
Comment (russelldj) [inline vital/types/rotation.cxx:62]: If you could point to your references for these calculations that would be helpful. But I'm willing to trust these are correct and just merge it.

Comment (russelldj) [reply vital/types/rotation.cxx:62]: I think pointing to a reference in a comment would be useful anyway, since this is a bit challenging to parse.

**Comment by mwoehlke-kitware** [comment]
> Can you explain why `static_cast` is preferred in this case?


I meant to leave a comment, looks like I got distracted...

If you're comparing to `T( 1 )`... because cast constructions are weird? 
Because you can search for a `static_cast` much more readily? If you're 
comparing to my alternate suggestion... meh, not a strong preference. I 
tend to prefer `auto ... = T{ ... }` vs. `T ... = ...`, but brace 
initialization in this context may throw a narrowing error. OTOH, `T{ 1 
}` would also be fine.

OTOH...

> I think my concern was if `T` was not a `double` or `float` and instead was an integer type. But that's probably pedantic and wouldn't happen in practice.

...if it's because you're worried about non-floating-point types, I 
wouldn't lose much sleep over it; I can't see but how a `rotation` 
wouldn't be horribly broken with those 🙂. (Of course, you can also just 
drop the `.0` if it makes you feel better. The actual suggested code is 
less my preference and more the first thing that came to mind.)

File: vital/types/rotation.cxx
Comment (daniel-riehm) [reply vital/types/rotation.cxx:63]: Done

**Comment by russelldj** [comment]
Everything here looks good to me. It sounds like the primary outstanding concern is whether the changed convention of the yaw, pitch, roll rotation constructor would have implications to other applications such as TeleSculptor.

**Comment by daniel-riehm** [comment]
I'm not sure exactly what form of verification would be satisfactory, but from my text search of the Telesculptor repo, it does not appear to use either the modified constructor nor the YPR conversion function in any `.cpp` or `.h` files.

**Comment by russelldj** [comment]
That's sufficient for me but I feel this decision is above my pay grade. Barring any immediate need to land it, I'd say wait on approval from someone else.

**Comment by russelldj** [comment]
@chetnieter @borovik135 Do you have any insight about whether changing the rotation constructor or YPR conversion functions might affect TeleSculptor? I know there's a good bit to get up to speed on here, so no need to look into this if it's not immediately obvious.

File: arrows/core/derive_metadata.cxx
Comment (mleotta) [reply arrows/core/derive_metadata.cxx:237]: It probably would have been best to make bug fixes for calculations in a separate branch from refactoring yaw-pitch-roll utility functions.  This would have allowed the simple fixes to land right away without getting held up by more complex changes that require more review.  However, no need to refactor PRs at this point.

**Review by mleotta — CHANGES_REQUESTED** [review]
These changes are great.  This is a much less kludgy way to handle YPR conversion from metadata.  My only concerns are that we should have a test for the ENU/NED conversions in the C++ tests (not just Python).  That is, here:
vital/tests/test_rotation.cxx

Also, I'm not entirely happy about `uas_ypr_to_rotation` being defined in `rotation.h`.  It seems a bit too application specific to be attached to the generic rotation class.  However, I don't have a good suggestion about where else to put it, so it's okay to stay here for now.

**Review by mleotta — CHANGES_REQUESTED** [review]
The new tests look good.  One more minor request.  You should add a brief release notes comment about this change in the `doc/release_notes/master.txt` file, especially since this is an API change that would impact users.  The `master.txt` file was recently reset with the recent KWIVER release, so you will need to rebase this branch on master or merger in master to get the updated release notes file before modifying.

**Review by mleotta — APPROVED** [review]
Will merge after CI passes

**Pull request merged**
